Objective C: use C++ exceptions on MinGW+GNUstep

qmfrederik opened the pull request:
The GNUstep Objective C runtime (libobjc2) is adding support for the GNU ABI on Windows (more specifically, MinGW).  The libobjc2 runtime uses C++ exceptions in that configuration; this PR updates clang to act accordingly.

The corresponding change to libobjc2 is here: https://github.com/gnustep/libobjc2/pull/267

qmfrederik committed:
Objective C: use C++ exceptions on MinGW+GNUstep

The GNUstep Objective C runtime (libobjc2) is adding support for
MinGW.  This runtime uses C++ exceptions in that configuration.

qmfrederik commented on clang/lib/CodeGen/CGException.cpp:
```diff
@@ -156,7 +154,11 @@ static const EHPersonality &getObjCPersonality(const TargetInfo &Target,
     return EHPersonality::NeXT_ObjC;
   case ObjCRuntime::GNUstep:
-    if (L.ObjCRuntime.getVersion() >= VersionTuple(1, 7))
+    if (T.isOSCygMing())
```
I moved the check for `isWindowsMSVCEnvironment` inside the `GNUstep` case, as on Windows we'll be using `MSVC_CxxFrameHandler3` when compiling with GNUstep + msvc; `GNU_CPlusPlus_SEH` when compiling with GNUstep + mingw and `GNU_ObjC_SEH` when compiling iwth GCC + mingw.

qmfrederik replied:
OK - this broke some tests, so I'll revert that.

qmfrederik commented:
/cc @davidchisnall

davidchisnall commented on clang/lib/CodeGen/CGObjCGNU.cpp:
```diff
@@ -2210,7 +2219,12 @@ CGObjCGNU::CGObjCGNU(CodeGenModule &cgm, unsigned runtimeABIVersion,
+  if ((CGM.getTarget().getTriple().isOSCygMing() &&
+       isRuntime(ObjCRuntime::GNUstep, 2))) {
+    ExceptionReThrowFn.init(&CGM, "objc_exception_rethrow", VoidTy, IdTy);
+  } else {
```
It looks as if both branches of the if and else here do the same thing?

qmfrederik replied:
On MinGW + objc2, `objc_exception_rethrow` is called; on other targets, `objc_exception_throw` is used (**re**throw vs throw).  I find it odd we were using `objc_exception_throw` to *re*throw an exception but decided to change it for MinGW+objc2 only.

davidchisnall replied:
Ah, sorry, I didn't see the `re` bit.

compnerd commented on clang/lib/CodeGen/CGException.cpp:
```diff
@@ -210,7 +212,10 @@ static const EHPersonality &getObjCXXPersonality(const TargetInfo &Target,
+    if (Target.getTriple().isOSCygMing())
+      return EHPersonality::GNU_CPlusPlus_SEH;
+    else
+      return EHPersonality::GNU_ObjCXX;
```
This might be nicer as a ternary return.

compnerd commented on clang/lib/CodeGen/CGObjCGNU.cpp:
```diff
@@ -819,10 +819,19 @@ class CGObjCGNUstep : public CGObjCGNU {
+        // void objc_exception_rethrow(void*)
+        ExceptionReThrowFn.init(&CGM, "__cxa_rethrow", PtrTy);
+      } else if (usesSEHExceptions) {
+        llvm::Type *VoidTy = llvm::Type::getVoidTy(VMContext);
```
Why not hoist this?  It is used across multiple branches.

qmfrederik replied:
Good remark, thanks.  Fixed this.

compnerd commented on clang/lib/CodeGen/CGObjCGNU.cpp:
```diff
@@ -3993,7 +4009,9 @@ void CGObjCGNU::EmitThrowStmt(CodeGenFunction &CGF,
-  if (isRethrow && usesSEHExceptions) {
+  if (isRethrow &&
+      (usesSEHExceptions || (CGM.getTarget().getTriple().isOSCygMing() &&
+                             isRuntime(ObjCRuntime::GNUstep, 2)))) {
```
Does it make sense to introduce a `usesCxxExceptions` variable?

qmfrederik replied:
Yes, that makes the intent clearer.  I changed this.

qmfrederik commented:
@compnerd Thanks for you feedback, I addressed your comments.

qmfrederik commented:
@compnerd @davidchisnall I've addressed the feedback and CI is (mostly) green.  The build failure on Windows seems unrelated:


```
# Removing c:/ws\src
# Creating "c:/ws\src"
> git clone -v -- https://github.com/llvm-premerge-tests/llvm-project.git .
Cloning into '.'...
remote: Internal Server Error
fatal: unable to access 'https://github.com/llvm-premerge-tests/llvm-project.git/': The requested URL returned error: 500
```

davidchisnall commented:
I'd like CI to be green.  Hopefully the failure is transient, but I don't seem to be able to kick off a rerun.  Maybe once others start passing, you can squash the two commits and force push to restart it?

qmfrederik commented:
Looks like there was an intermittent issue with GitHub, which should be resolved now.  I've squashed and pushed.  Fingers crossed!

qmfrederik commented:
CI looks green, let me know if there's anything else you need from me on this PR!

compnerd commented on clang/lib/CodeGen/CGObjCGNU.cpp:
```diff
@@ -2210,7 +2219,11 @@ CGObjCGNU::CGObjCGNU(CodeGenModule &cgm, unsigned runtimeABIVersion,
   ExceptionThrowFn.init(&CGM, "objc_exception_throw", VoidTy, IdTy);
-  ExceptionReThrowFn.init(&CGM, "objc_exception_throw", VoidTy, IdTy);
+  if (usesCxxExceptions) {
+    ExceptionReThrowFn.init(&CGM, "objc_exception_rethrow", VoidTy, IdTy);
```
I would say that a ternary to select between the exception model and identify the name would be better as you are just selecting the label.

davidchisnall commented:
It looks as if three tests are failing in CI:

  Clang :: CodeGenObjC/2007-04-03-ObjcEH.m
  Clang :: CodeGenObjC/exceptions-personality.m
  Clang :: Coverage/codegen-gnu.m
